File: src/testWithSpringBoot_1_5/java/org/openrewrite/java/spring/boot2/MigrateLocalServerPortAnnotationTest.java
/*
 * Copyright 2021 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java.spring.boot2;

import org.junit.jupiter.api.Test;
import org.openrewrite.Issue;
import org.openrewrite.java.JavaParser;
import org.openrewrite.test.RecipeSpec;
import org.openrewrite.test.RewriteTest;

import static org.openrewrite.java.Assertions.*;
import static org.openrewrite.maven.Assertions.pomXml;

class MigrateLocalServerPortAnnotationTest implements RewriteTest {
    @Override
    public void defaults(RecipeSpec spec) {
        spec.recipe(new MigrateLocalServerPortAnnotation())
          .parser(JavaParser.fromJavaVersion().classpath("spring-boot"));
    }

    @Issue("https://github.com/openrewrite/rewrite-spring/issues/116")
    @Test
    void shouldReplaceType() {
        //language=java
        rewriteRun(
          java(
            """
              import org.springframework.boot.context.embedded.LocalServerPort;

              public class RandomTestClass {
                  @LocalServerPort
                  private int port;
              }
              """,
            """
              import org.springframework.boot.web.server.LocalServerPort;

              public class RandomTestClass {
                  @LocalServerPort
                  private int port;
              }
              """
          )
        );
    }

    @Issue("https://github.com/openrewrite/rewrite-spring/issues/541")
    @Test
    void shouldAddDependencyCorrectly() {
        rewriteRun(
          spec -> spec.expectedCyclesThatMakeChanges(2),
          mavenProject("project",
            srcMainJava(
              //language=Java
              java(
                """
                  import org.springframework.boot.context.embedded.LocalServerPort;

                  class RandomTestClass {
                      @LocalServerPort
                      private int port;
                  }
                  """,
                """
                  import org.springframework.boot.web.server.LocalServerPort;

                  class RandomTestClass {
                      @LocalServerPort
                      private int port;
                  }
                  """
              )
            ),
            //language=XML
            pomXml(
              """
                <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
                    <modelVersion>4.0.0</modelVersion>
                    <parent>
                        <groupId>org.springframework.boot</groupId>
                        <artifactId>spring-boot-starter-parent</artifactId>
                        <version>2.0.9.RELEASE</version>
                        <relativePath/>
                    </parent>
                    <groupId>com.example</groupId>
                    <artifactId>acme</artifactId>
                    <version>0.0.1-SNAPSHOT</version>
                    <dependencies>
                        <dependency>
                            <groupId>org.springframework.boot</groupId>
                            <artifactId>spring-boot-starter</artifactId>
                        </dependency>
                    </dependencies>
                </project>
                """,
              """
                <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
                    <modelVersion>4.0.0</modelVersion>
                    <parent>
                        <groupId>org.springframework.boot</groupId>
                        <artifactId>spring-boot-starter-parent</artifactId>
                        <version>2.0.9.RELEASE</version>
                        <relativePath/>
                    </parent>
                    <groupId>com.example</groupId>
                    <artifactId>acme</artifactId>
                    <version>0.0.1-SNAPSHOT</version>
                    <dependencies>
                        <dependency>
                            <groupId>org.springframework.boot</groupId>
                            <artifactId>spring-boot-starter</artifactId>
                        </dependency>
                        <dependency>
                            <groupId>org.springframework.boot</groupId>
                            <artifactId>spring-boot-starter-web</artifactId>
                        </dependency>
                    </dependencies>
                </project>
                """
            )
          )
        );
    }
}

File: src/testWithSpringBoot_1_5/kotlin/org/openrewrite/java/spring/boot2/MigrateLocalServerPortAnnotationTest.kt
/*
 * Copyright 2021 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java.spring.boot2

import org.junit.jupiter.api.Test
import org.openrewrite.Issue
import org.openrewrite.Recipe
import org.openrewrite.java.JavaParser
import org.openrewrite.java.JavaRecipeTest

class MigrateLocalServerPortAnnotationTest : JavaRecipeTest {
    override val parser: JavaParser
        get() = JavaParser.fromJavaVersion()
            .classpath("spring-boot")
            .build()

    override val recipe: Recipe
        get() = MigrateLocalServerPortAnnotation()

    @Issue("https://github.com/openrewrite/rewrite-spring/issues/116")
    @Test
    fun givenHasStringVariableWhenRemovingDeprecatedThenReplacesAddEnvironmentWithSetProperties() = assertChanged(
        before = """
            import org.springframework.boot.context.embedded.LocalServerPort;

            public class RandomTestClass {
                @LocalServerPort
                private int port;
            }
        """,
        after = """
            import org.springframework.boot.web.server.LocalServerPort;
            
            public class RandomTestClass {
                @LocalServerPort
                private int port;
            }
        """
    )
}
